### PR TITLE
Diagnostics report tool, first iteration

### PR DESCRIPTION
**Pull request by klaren**
This introduces `neo4j-admin report` that can be used to collect information useful for support cases.

The first iteration includes a way to gather files offline, for startup troubleshooting, and a limited number of online information that can be accessed through the standard JMX interface.

You can pass `--list` to the command to get a view of available classifiers, e.g.
```
/bin> ./neo4j-admin report --list
Trying to find running instance of neo4j
Attached to running process with process id 6439
Connected to JMX endpoint
All available classifiers:
  config     include configuration file
  logs       include log files
  plugins    include a view of the plugin directory
  sysprop    include a list of java system properties
  threads    include a thread dump of the running instance
  tree       include a view of the tree structure of the data directory
```

A sample output follows:
```
/bin> ./neo4j-admin report all   
Trying to find running instance of neo4j
Attached to running process with process id 6439
Connected to JMX endpoint
Writing report to /home/klaren/Desktop/neo4j-community-3.4.0-SNAPSHOT/bin/reports/2017-12-14_132858.zip
     1/7 [####################] 100%  logs/debug.log
     2/7 [####################] 100%  logs/neo4j.log
     3/7 [####################] 100%  plugins.txt
     4/7 [####################] 100%  tree.txt
     5/7 [####################] 100%  vm.prop
     6/7 [####################] 100%  threaddump.txt
     7/7 [####################] 100%  neo4j.conf
```

  Known issues:
-----------------------
* Online sources does not work on Windows due to the way we start neo4j with `prunsrv`

### CODE REVIEW

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandProvider.java
Comment (tinwelint) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandProvider.java]: Would be nice if the default list would be an accessible constant which this string would include, to not let them go out of sync in the future.

**Comment by tinwelint** [comment]
And generally regarding errors, it may be nice having the ability to print stack traces... We all know how frustrating it is to try and figure out a cause by only seeing the error.toString()

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: a bit more fields can be final in this class

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: lets extract pid reading into separate helper method

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java:97]: a lot of things is going on inside this method would be good to extract some parts out of it

File: community/kernel/src/main/java/org/neo4j/diagnostics/ProgressAwareInputStream.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/diagnostics/ProgressAwareInputStream.java:82]: i would add additional check to make sure we will never make it to more then 100 :)

File: community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java]: this stream needs to be closed

File: community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java]: would be nice to close this connector in the end

File: community/dbms/src/test/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandIT.java
Comment (MishaDemianenko) [inline community/dbms/src/test/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandIT.java:114]: we need to check that this works on all vm's that we need to support

File: community/dbms/src/test/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandTest.java
Comment (MishaDemianenko) [inline community/dbms/src/test/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandTest.java]: we need to restore previous property value in tear down

File: community/dbms/pom.xml
Comment (MishaDemianenko) [inline community/dbms/pom.xml:51]: wdyt about moving this profiles to root pom instead?

File: community/dbms/pom.xml
Comment (MishaDemianenko) [inline community/dbms/pom.xml]: something to cleanup

File: community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReporter.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReporter.java]: 🚳

File: community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReportSources.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReportSources.java]: this stream needs to be closed

File: community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReportSources.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReportSources.java]: StandardCharsets.UTF_8

**Review by MishaDemianenko — CHANGES_REQUESTED** [review]

**Comment by MishaDemianenko** [comment]
@klaren couple of comments also i think we need to think about moving it out of kernel as much as possible and maybe move all of the diagnostics inside dbms module into its own package

File: community/dbms/pom.xml
Comment (klaren) [reply community/dbms/pom.xml:51]: I feel horrible just including this, let's keep the exposure to a minimum.

File: community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java
Comment (klaren) [reply community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java]: In the current state, this will always use RMI and that does not require a proper close call.

File: community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java
Comment (klaren) [reply community/dbms/src/main/java/org/neo4j/dbms/report/jmx/JmxDump.java]: Nice catch!

**Comment by klaren** [comment]
@MishaDemianenko I tried to reduce the code in kernel, the problem is
```java
import org.neo4j.kernel.configuration.Config;
```
that gives a cycle in the dependencies.

File: community/dbms/src/test/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandIT.java
Comment (klaren) [reply community/dbms/src/test/java/org/neo4j/commandline/dbms/DiagnosticsReportCommandIT.java:114]: Yes, should probably change that assume to assert.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/diagnostics/ClusterDiagnosticsOfflineReportProvider.java
Comment (MishaDemianenko) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/diagnostics/ClusterDiagnosticsOfflineReportProvider.java]: it can be a good idea to supply filename filter instead to 'fs.listFiles' as second argument

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/diagnostics/ClusterDiagnosticsOfflineReportProvider.java
Comment (MishaDemianenko) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/diagnostics/ClusterDiagnosticsOfflineReportProvider.java]: File.separator instead of "/"?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: for some of the users `/tmp` can be a mysterious example, maybe we can change it to something more verbose/descriptive that also use platform base file.separator?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: please use isEmpty instead

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: lets convert those todos into cards

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: maybe we can convert orphans to set directly where we initialize it?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: will `StringEscapeUtils.escapeCsv( s )` work instead?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: would suggest asking our cs eng. about names for those classifiers if they think its understandable for them and for customers

File: community/dbms/src/main/java/org/neo4j/dbms/diagnostics/jmx/JmxDump.java
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/dbms/diagnostics/jmx/JmxDump.java]: todo

File: community/dbms/pom.xml
Comment (MishaDemianenko) [inline community/dbms/pom.xml]: we do not really need this profile right?

File: community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReporter.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReporter.java]: todo

**Review by MishaDemianenko — CHANGES_REQUESTED** [review]
Couple of code comments.
Also even for first iterations we need to have a bit more test at least around code reporting code (progress is reported, report presented to user etc)
We need to ask about what kind of documentation needs to be updated and operability team should take a look for our entry script/naming/etc.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java
Comment (klaren) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/DiagnosticsReportCommand.java]: yes it will!

**Review by MishaDemianenko — APPROVED** [review]

**Comment by MishaDemianenko** [comment]
@klaren approved, but please not failing test that fails because of the unpredictable order of classifiers.

**Review by praveenag — APPROVED** [review]
Approved once the changes requested are applied and merged with this PR

**Pull request merged**
